File: proj.android_studio/ee-x-firebase-performance/src/main/java/com/ee/firebase/performance/FirebasePerformanceTrace.java
package com.ee.firebase.performance;

import androidx.annotation.NonNull;

import com.ee.core.IMessageBridge;
import com.ee.core.internal.JsonUtils;
import com.google.firebase.perf.metrics.Trace;

import java.util.Map;

public class FirebasePerformanceTrace {
    private static final String k__key = "key";
    private static final String k__value = "value";

    private IMessageBridge _bridge;
    private String _traceName;
    private Trace _trace;

    public FirebasePerformanceTrace(@NonNull IMessageBridge bridge,
                                    @NonNull Trace trace,
                                    @NonNull String traceName) {
        _bridge = bridge;
        _traceName = traceName;
        _trace = trace;
        registerHandlers();
    }

    void destroy() {
        deregisterHandlers();
        _bridge = null;
        _trace = null;
        _traceName = null;
    }

    @NonNull
    private String k__start() {
        return "FirebasePerformance_start_" + _traceName;
    }

    @NonNull
    private String k__stop() {
        return "FirebasePerformance_stop_" + _traceName;
    }

    @NonNull
    String k__incrementMetric() {
        return "FirebasePerformance_incrementMetric_" + _traceName;
    }

    @NonNull
    String k__getLongMetric() {
        return "FirebasePerformance_getLongMetric_" + _traceName;
    }

    @NonNull
    String k__putMetric() {
        return "FirebasePerformance_putMetric_" + _traceName;
    }

    private void registerHandlers() {
        _bridge.registerHandler(message -> {
            start();
            return "";
        }, k__start());

        _bridge.registerHandler(message -> {
            stop();
            return "";
        }, k__stop());

        _bridge.registerHandler(message -> {
            Map<String, Object> dict = JsonUtils.convertStringToDictionary(message);
            assert dict != null;

            String key = (String) dict.get(k__key);
            Integer value = (Integer) dict.get(k__value);
            incrementMetric(key, value);
            return "";
        }, k__incrementMetric());

        _bridge.registerHandler(message -> {
            Map<String, Object> dict = JsonUtils.convertStringToDictionary(message);
            assert dict != null;

            String key = (String) dict.get(k__key);
            Long value = (Long) dict.get(k__value);
            putMetric(key, value);
            return "";
        }, k__putMetric());

        _bridge.registerHandler(message -> {
            String metricName = message;
            return String.valueOf(getLongMetric(metricName));
        }, k__getLongMetric());
    }

    private void deregisterHandlers() {
        _bridge.deregisterHandler(k__start());
        _bridge.deregisterHandler(k__stop());
        _bridge.deregisterHandler(k__incrementMetric());
        _bridge.deregisterHandler(k__putMetric());
        _bridge.deregisterHandler(k__getLongMetric());
    }

    public void start() {
        _trace.start();
    }

    public void stop() {
        _trace.stop();
    }

    public void incrementMetric(@NonNull String traceName, long value) {
        _trace.incrementMetric(traceName, value);
    }

    public void putMetric(@NonNull String name, long value) {
        _trace.putMetric(name, value);
    }

    public long getLongMetric(@NonNull String name) {
        return _trace.getLongMetric(name);
    }
}

File: proj.android_studio/ee-x-firebase-performance/src/main/java/com/ee/firebase/performance/FirebasePerformanceTrace.kt
package com.ee.firebase.performance

import androidx.annotation.AnyThread
import com.ee.core.IMessageBridge
import com.ee.core.internal.deserialize
import com.ee.core.registerHandler
import com.google.firebase.perf.metrics.Trace
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.UnstableDefault

@ImplicitReflectionSerializer
@UnstableDefault
internal class FirebasePerformanceTrace(
    private val _bridge: IMessageBridge,
    private val _trace: Trace,
    private val _traceName: String) {
    init {
        registerHandlers()
    }

    fun destroy() {
        deregisterHandlers()
    }

    private val k__start: String
        get() = "FirebasePerformance_start_$_traceName"

    private val k__stop: String
        get() = "FirebasePerformance_stop_$_traceName"

    private val k__incrementMetric: String
        get() = "FirebasePerformance_incrementMetric_$_traceName"

    private val k__getLongMetric: String
        get() = "FirebasePerformance_getLongMetric_$_traceName"

    private val k__putMetric: String
        get() = "FirebasePerformance_putMetric_$_traceName"

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerHandler(k__start) {
            start()
            ""
        }
        _bridge.registerHandler(k__stop) {
            stop()
            ""
        }
        _bridge.registerHandler(k__incrementMetric) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: Long
            )

            val request = deserialize<Request>(message)
            incrementMetric(request.key, request.value)
            ""
        }
        _bridge.registerHandler(k__putMetric) { message ->
            @Serializable
            class Request(
                val key: String,
                val value: Long
            )

            val request = deserialize<Request>(message)
            putMetric(request.key, request.value)
            ""
        }
        _bridge.registerHandler(k__getLongMetric) { message ->
            getLongMetric(message).toString()
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(k__start)
        _bridge.deregisterHandler(k__stop)
        _bridge.deregisterHandler(k__incrementMetric)
        _bridge.deregisterHandler(k__putMetric)
        _bridge.deregisterHandler(k__getLongMetric)
    }

    @AnyThread
    fun start() {
        _trace.start()
    }

    @AnyThread
    fun stop() {
        _trace.stop()
    }

    @AnyThread
    private fun incrementMetric(traceName: String, value: Long) {
        _trace.incrementMetric(traceName, value)
    }

    @AnyThread
    private fun putMetric(name: String, value: Long) {
        _trace.putMetric(name, value)
    }

    @AnyThread
    private fun getLongMetric(name: String): Long {
        return _trace.getLongMetric(name)
    }
}